### ISSUE TITLE
CPU Spike & WriteRedisConnectionException on Elasticache Redis Data Tiering with Redisson

### ISSUE THREAD

**Issue by avhirupc** [redisson/redisson#4179]
Elastic-cache recently released a [data-tiering](https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/data-tiering.html) which lets offload LRU keys to disk.

In an attempt to onboard to the feature, we ran some tests where we would fill up our memory till 100% and observe keys being offloaded to disk. We repeated the test using both redis-benchmark tool & Redisson ( Client we use for production). 

With redisson as soon as MemoryUtilisation reached 100% we observed a sudden spike in CPU , accompanied with `WriteRedisConnectionException` . This CPU spike lead to node failures again and again. We didnt observe the same behaviour with redis-benchmark tool. 

My question is : Why would there be a spike in CPU with Redisson ? How can we avoid this & is there an ability to perform exponential backoff timeout in redisson?

Redisson Error Message when memory reaches 100%  : `Error message: org.redisson.client.WriteRedisConnectionException: Unable to write command into connection! Increase connection pool size).`

** Setup for Replicating behaviour ** 
Redis Cluster : 1 shard with 1 replica of node type r6gd.2xlarge.
Redisson Config : 

```
REDIS_SCAN_INTERVAL_MS = (int) Duration.ofMinutes(1).toMillis();
REDIS_PING_CONNECTION_INTERVAL_MS = 5000;
RETRY_INTERVAL = 50;
RETRY_ATTEMPTS = 3;
TIMEOUT = 30000;
CONNECTION_TIMEOUT = 60000 * 5;
NETTY_THREADS = 128;

redissonConfig
        .useClusterServers()
        .addNodeAddress(redisAddress)
        .setScanInterval(REDIS_SCAN_INTERVAL_MS)
        .setRetryInterval(RETRY_INTERVAL)
        .setRetryAttempts(RETRY_ATTEMPTS)
        .setPingConnectionInterval(REDIS_PING_CONNECTION_INTERVAL_MS)
        .setKeepAlive(true)
        .setConnectTimeout(CONNECTION_TIMEOUT)
        .setTimeout(TIMEOUT);
```

 IMO, the timeouts are already quite high. I dont see new connections being formed as well.

**Comment by mrniko** [comment]
Can you share test code for Redisson?

**Comment by mrniko** [comment]
Did you try to increase RETRY_INTERVAL ?

**Comment by avhirupc** [comment]
Yes , IIRC I did try it out with increasing the RETRY_INTERVAL to 100.

Here is a dummy code on what how we are ingesting
```java
import org.redisson.api.BatchOptions;
import lombok.Value;
import org.redisson.codec.SnappyCodecV2;
import org.redisson.api.BatchResult;
import org.redisson.api.RBatchReactive;
import org.redisson.api.RMapReactive;
import org.redisson.client.codec.Codec;
import java.util.stream.Stream;

@Value
pubic class Row {
    String HashKey;
    String MapKey;
    String Value;
}

public class DummyIngest {
    
    public long ingest(Stream<Row> inputStream, long timeToLive) {
    log.debug("Starting bulk insert to Redis...");

    Codec codec = new SnappyCodecV2();
    int SET_BATCH_SIZE = 500;
    final Long linesObj = Flux.fromStream(inputStream)
            .buffer(SET_BATCH_SIZE)
            .flatMap(rowList -> {
                RBatchReactive batch = client.createBatch(batchOptions);
                rowList.stream()
                        .map(kv -> {
                            String hashKey = kv.getHashKey();
                            String mapKey = kv.getMapKey();
                            batch.getMap(hashKey, codec)
                                    .fastPut(mapKey, kv.getValue());
                            return hashKey;
                        })
                        .distinct()
                        .forEach(hash -> batch.getKeys().expire(hash, timeToLive, TimeUnit.MINUTES));
                return batch.execute().thenReturn(rowList.size());
            })
            .reduce(0L, Long::sum)
            .block();

    long lines = linesObj == null ? 0L : linesObj;
    log.debug("Sent {} more records to Redis.", lines);
    return lines;
    }
    
    public static void main(String args[]) {
        // Have a File with three columns : hashKey, mapKey, & Value
        // Read file & create stream 
        // Call Ingest Method`
        
    }


}
```

**Comment by mrniko** [comment]
> With redisson as soon as MemoryUtilisation reached 100% we observed a sudden spike in CPU

Is there CPU spike on Redis side during that moment?

**Comment by avhirupc** [comment]
Yes, I mean 100% Redis Memory utilization. The CPU spike is on Redis side. The CPU spike only happens when using Redisson Client.  In summary, 100% Redis Memory utilization correlates with spike in CPU to 100% with redisson client. 

Chatting with Elasticache engineer, We didn't observe anything strange happening on Redis cluster side except CPU spike & node failure.

One observation I made : The keys were pushed to disk as expected when we reached 100% memory utilisation. Shortly after we saw a spike in NewConnections being formed, which is when the node failure happens. 

In summary there are two questions . 
1. Why would the CPU Spike to 100% when start pushing keys to disk.
2. Why is a new pool of connections being formed ?

Let me know if you need something to help you replicate the behaviour.

**Comment by mrniko** [comment]
> Shortly after we saw a spike in NewConnections being formed, which is when the node failure happens.

This happened since during this operation Redis connections where busy with command execution. Default minimumIdleSize is 24 and poolSize = 64 for Redis nodes, so there is a room to grow if current connections are busy. Which we observe. To avoid this you can set minimumIdleSize = poolSize = 32 or any other value.

> Why would the CPU Spike to 100% when start pushing keys to disk.

It's better to ask AWS Elasticache Team.

**Comment by avhirupc** [comment]
Thanks for giving insight on why new connections were being formed. 

I did have multiple deep dives with Elasticache Team. Using a different client(redis-benchmark) we didnt observe same behaviour. Also , they tested it out with their clients and everything worked fined with no spike. 

CPU spike when memory reaches 100% seems to be only happen when using Redisson. So it would be helpful to dig into this, as I assume soon a lot of people would be onboarding to data-tiering.

**Comment by mrniko** [comment]
Can you try it again with fixed pool size for slave and master nodes?

**Comment by avhirupc** [comment]
Sure. I can, I ll reach back after trying that out.

**Comment by mrniko** [comment]
Have you resolved this issue?
